alphacsc/_encoder.py

```diff
@@ -148,13 +148,43 @@ def get_z_encoder_for(
             n_jobs,
             solver_kwargs,
             algorithm,
-            reg
+            reg,
+            loss
         )
     else:
         raise ValueError(f'unrecognized solver type: {solver}.')
 
 
 class BaseZEncoder:
+
+    def __init__(
+            self,
+            X,
+            D_hat,
+            n_atoms,
+            atom_support,
+            n_jobs,
+            solver_kwargs,
+            algorithm,
+            reg,
+            loss):
+
+        self.X = X
+        self.D_hat = D_hat
+        self.n_atoms = n_atoms
+        self.atom_support = atom_support
+        self.n_jobs = n_jobs
+
+        self.solver_kwargs = solver_kwargs
+        self.algorithm = algorithm
+        self.reg = reg
+        self.loss = loss
+
+        self.n_trials, self.n_channels, self.n_times = X.shape
```

Comment:
I would not store self.n_trials, self.n_channels, self.n_times as it's available in self.X
to make sure the self.X and self.n_trials, self.n_channels, self.n_times are always in sync
I would consider properties but for now just do:

n_trials, n_channels, n_times = self.X

when you need it